File: jambeth/jAmbeth.IoC/src/de/osthus/ambeth/ioc/IocBootstrapModule.java
package de.osthus.ambeth.ioc;

import de.osthus.ambeth.ioc.annotation.FrameworkModule;
import de.osthus.ambeth.ioc.factory.IBeanContextFactory;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LogInstance;
import de.osthus.ambeth.proxy.CgLibUtil;
import de.osthus.ambeth.proxy.ICgLibUtil;
import de.osthus.ambeth.threading.GuiThreadHelper;
import de.osthus.ambeth.threading.IGuiThreadHelper;
import de.osthus.ambeth.util.DedicatedConverterUtil;
import de.osthus.ambeth.util.DelegatingConversionHelper;
import de.osthus.ambeth.util.IConversionHelper;
import de.osthus.ambeth.util.IDedicatedConverterExtendable;
import de.osthus.ambeth.util.IInterningFeature;
import de.osthus.ambeth.util.InterningFeature;
import de.osthus.ambeth.util.converter.BooleanArrayConverter;
import de.osthus.ambeth.util.converter.ByteArrayConverter;
import de.osthus.ambeth.util.converter.CharArrayConverter;

@FrameworkModule
public class IocBootstrapModule implements IInitializingModule
{
	@SuppressWarnings("unused")
	@LogInstance
	private ILogger log;

	@Override
	public void afterPropertiesSet(IBeanContextFactory beanContextFactory) throws Throwable
	{
		beanContextFactory.registerBean("conversionHelper", DelegatingConversionHelper.class).propertyRefs("*conversionHelper")
				.autowireable(IConversionHelper.class, IDedicatedConverterExtendable.class);

		beanContextFactory.registerBean("booleanArrayConverter", BooleanArrayConverter.class);
		DedicatedConverterUtil.biLink(beanContextFactory, "booleanArrayConverter", String.class, boolean[].class);

		beanContextFactory.registerBean("byteArrayConverter", ByteArrayConverter.class);
		DedicatedConverterUtil.biLink(beanContextFactory, "byteArrayConverter", String.class, byte[].class);

		beanContextFactory.registerBean("charArrayConverter", CharArrayConverter.class);
		DedicatedConverterUtil.biLink(beanContextFactory, "charArrayConverter", String.class, char[].class);

		beanContextFactory.registerBean("interningFeature", InterningFeature.class).autowireable(IInterningFeature.class);

		beanContextFactory.registerBean("cgLibUtil", CgLibUtil.class).autowireable(ICgLibUtil.class);

		beanContextFactory.registerBean("guiThreadHelper", GuiThreadHelper.class).autowireable(IGuiThreadHelper.class);
	}
}

File: ambeth/Ambeth.IoC/ioc/IocBootstrapModule.cs
using De.Osthus.Ambeth.Ioc.Annotation;
using De.Osthus.Ambeth.Ioc.Factory;
using De.Osthus.Ambeth.Log;
using De.Osthus.Ambeth.Threading;
using De.Osthus.Ambeth.Util;
using De.Osthus.Ambeth.Util.Converter;
using System;

namespace De.Osthus.Ambeth.Ioc
{
    [FrameworkModule]
    public class IocBootstrapModule : IInitializingBootstrapModule
    {
        [LogInstance]
		public ILogger Log { private get; set; }
                
        public void AfterPropertiesSet(IBeanContextFactory beanContextFactory)
        {
            beanContextFactory.RegisterBean<ThreadPool>("threadPool").Autowireable(typeof(IThreadPool), typeof(IDelayedExecution));

            beanContextFactory.RegisterBean<DelegatingConversionHelper>("conversionHelper").PropertyRefs("*conversionHelper")
				.Autowireable(typeof(IConversionHelper), typeof(IDedicatedConverterExtendable));

		    beanContextFactory.RegisterBean<BooleanArrayConverter>("booleanArrayConverter");
            DedicatedConverterUtil.BiLink(beanContextFactory, "booleanArrayConverter", typeof(String), typeof(bool[]));

            beanContextFactory.RegisterBean<ByteArrayConverter>("byteArrayConverter");
            DedicatedConverterUtil.BiLink(beanContextFactory, "byteArrayConverter", typeof(String), typeof(byte[]));

            beanContextFactory.RegisterBean<SByteArrayConverter>("sbyteArrayConverter");
            DedicatedConverterUtil.BiLink(beanContextFactory, "sbyteArrayConverter", typeof(String), typeof(sbyte[]));
            
            beanContextFactory.RegisterBean<CharArrayConverter>("charArrayConverter");
            DedicatedConverterUtil.BiLink(beanContextFactory, "charArrayConverter", typeof(String), typeof(char[]));

            beanContextFactory.RegisterBean<InterningFeature>("interningFeature").Autowireable<IInterningFeature>();

            beanContextFactory.RegisterBean<GuiThreadHelper>("guiThreadHelper").Autowireable<IGuiThreadHelper>();
        }
    }
}
